src/ir/auth_logic/universe_relation_insertion.cc

```diff
@@ -0,0 +1,357 @@
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/ir/auth_logic/universe_relation_insertion.h"
+
+namespace raksha::ir::auth_logic {
+
+namespace {
+// Determine if an identifier (which may be an argument of a predicate
+// or a principal name) is a constant or variable.
+// TODO(#673 aferr) it would be better to refactor the AST so that
+// identifiers are a separate AST node with constant and non-constant
+// children determined at parse time rather than here since
+// this information is available at parse time.
+bool IsNameConstant(absl::string_view id) {
+  if (id[0] == '"') {
+    return true;
+  } else {
+    // Numeric literals are constants and this is
+    // a numeric literal if all the characters are numbers.
+    for (char c : id) {
+      if (!isdigit(c)) return false;
+    }
+    return true;
+  }
+}
+
+// (TODO: #672) once an AST node for numbers as RVALUEs in numeric
+// comparisons is added, this visitor should also visit
+// the numeric RVALUEs and add them to the type environment
+// with type ArgumentType(kNumber, "Number")
+// This is a workaround that makes use of the fact that
+// numeric comparisons are represented as predicates
+// with a name that matches the operator:
+// (https://github.com/google-research/raksha/blob/be6ef8e1e1a20735a06637c12db9ed0b87e3d2a2/src/ir/auth_logic/ast_construction.cc#L92)
+static inline bool PredicateIsNumericOperator(const datalog::Predicate& pred) {
+  if (pred.name() == "<" || pred.name() == "<" || pred.name() == "=" ||
+      pred.name() == "!=" || pred.name() == "<=" || pred.name() == ">=") {
```

Comment:
In general, it's best to not repeat a method call over and over when you expect it to always return the same thing. Just call it once and assign it to a temporary.

Also, `if (predicate) return true; else return false;` is equivalent to `return predicate`.

Also, I wonder whether it might be more readable to just create a set of the options and look this up in the set like so:

```C++
static inline bool PredicateIsNumericOperator(const datalog::Predicate &pred) {
  const absl::flat_hash_set numeric_operators = {
   "<", ...
  };
  return numeric_operators.contains(pred.name());
}
```

Also, I think you may have accidentally repeated `<` rather than put in `>`.